controllers/raffleController.js

```diff
@@ -91,8 +96,27 @@ raffleController.post(
   async (request, response) => {
     const r = request.body;
     try {
-      const newRow = await addNewParticipant(r, request.params.id);
-      if (newRow) response.status(201).json({ data: newRow });
+      const result = await getRaffleById(request.params.id);
+      if (!result) {
+        return response.status(404).json({
+          error: `Could not find raffle with id ${request.params.id}`,
+        });
+      } else if (result?.winner_id != null)
+        return response.status(400).json({
+          error: `raffle with id ${request.params.id} is not accepting new participants`,
+        });
+    } catch (error) {
+      return response.status(500).json({ error: error.message });
+    }
+
+    try {
+      await addNewParticipant(r);
+      const newRowJt = await addNewRaffleParticipant(request.params.id, r);
```

Comment:
I'm not sure this handles all of the possible cases yet. For example:
1. the participant already exists in the participants table. In this case we don't want to add them again.
2. the participant is supposed to exist in the participants table, but doesn't. In this case we should return a 404.
3. the participant is brand new and needs to be created in the participants table. (Your code handles this case, which is awesome)
4. the attempt to add the new participant _fails_ (for example because of a bug or because the db connection was lost). In this case we do not want to proceed and we do not want to insert anything in `raffles_participant`. (I think your code will probably handle this by catching the error.)
5. the attempt to add the new participant _succeeds_ but the _next_ query fails: inserting into `raffles_participant`. The `try/catch` will probably partly handle this but we have a problem because we should probably roll back to insert to the participants table, or at least make it very clear to the client that the request was only a partial success.

There are a few different ways to handle this. You could give the front end more work to do, like have it first attempt to create a participant then, if successful, make a second request to add the participant to the raffle. This handles the rollback problem because each API call results in only one database query.

A second approach is to put the two insert queries into a `transaction`, a way of telling postgres that you want either all of the queries to succeed or none, but not some. pg promise supports transactions so you can do this.

Either way, I haven't looked at your front end but I think it would be good if your front end fetched the participants table before any of this happened. That way the front end will know if the client is adding an existing participant or if we need to create a new participant and add them to a raffle.

This is what I meant when I said the join table expands the scope and complexity a lot!

Reply:
Thanks I like your feedback. Many of the things you mentioned I have considered. I do like the idea of getting all participants in a raffle and having that information ready before filling a new participant form.  I thought this would not quite be ideal because a raffle could already draw a winner before the form is submitted, and it would be generally harmless to keep a table of users of the app regardless if they successfully bought tickets.